Meson docs

samay-sharma opened the pull request.

samay-sharma commented:
Screenshots of the pages I'd like review on

Index structure

<img width="684" alt="Screen Shot 2022-08-10 at 12 20 43 PM" src="https://user-images.githubusercontent.com/6285167/184002755-08f7c42f-dfdf-4d98-a4f5-c229080011be.png">

Short version
<img width="1411" alt="Screen Shot 2022-08-10 at 12 20 56 PM" src="https://user-images.githubusercontent.com/6285167/184002838-236c40ff-4878-4221-a63f-2b486d5ec1ae.png">

samay-sharma commented:
Specific topics to discuss:

* Is debug (which is the default) the right level we want people to build in short version?
* Meson compile vs ninja - Any reason to prefer one over the other?
* Where to install and put data directory? (to avoid sudo)
 I currently put it in the postgres directory. So, the source folder will now have 3 additional folders: build, install and data. This may cause issues with git tracking install and data too. I also don't know if we should assume that we have access / permission to put files in a directory outside the downloaded source. So, want your thoughts on that.
* Running tests takes forever - should we have them in short version (current docs don't)?

eli-schwartz commented:
> Meson compile vs ninja - Any reason to prefer one over the other?

The former will run an extra python process which is a bit wasteful, but in return it can tell if you passed the `--backend` argument to `meson setup` and will make sure to run the correct backend, even if that is `msbuild` instead of `ninja`.

Since the short docs don't suggest anything about selecting a different backend, assuming that it is ninja and simply running `ninja` should be fine, and trivially faster.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,2635 @@
+ </para>
+
+ <para>
+  What is Meson and Why Meson?
```
I'm not sure that makes a lot of sense to have here?

samay-sharma replied:
Removed. But, I think we should have some content on why users should use one over the other OR why we recommend one over the other.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,2635 @@
+  If you are building <productname>PostgreSQL</productname> for Microsoft
+  Windows, read this chapter if you intend to build with MinGW or Cygwin;
+  but if you intend to build with Microsoft's <productname>Visual
+  C++</productname>, see <xref linkend="install-windows"/> instead.
```
That's probably outdated now, right?

samay-sharma replied:
Removed.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,2635 @@
+# current instructions for testing (to be removed when merging)
+git clone https://github.com/anarazel/postgres.git
+cd postgres
+git switch meson
```
I'd add this to the clone as -b meson instead. Worth noting that git switch is relatively new, so plenty git versions won't have it yet.

samay-sharma replied:
Changed.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,2635 @@
+
+# setup and enter build directory (done only first time)
+# meson setup build
+meson setup build --prefix=$PWD/install
```
Using $PWD has the disadvantage of not working on windows. Wonder if we can come up with something better.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,2635 @@
+    <listitem>
+     <para>
+      <indexterm>
+       <primary>make</primary>
```
Needs updating...

samay-sharma replied:
Updated.

anarazel reviewed (commented):
Good start!

Think it'd be a bit easier to review if more of the non-applicable stuff were stripped out instead of copied...

anarazel commented:
> Running tests takes forever - should we have them in short version (current docs don't)?

Let's discuss that on the list

anarazel commented:
> Where to install and put data directory? (to avoid sudo)
> I currently put it in the postgres directory. So, the source folder will now have 3 additional folders: build, install and data. This may cause issues with git tracking install and data too. I also don't know if we should assume that we have access / permission to put files in a directory outside the downloaded source. So, want your thoughts on that.

We could just git clone into postgresql/src and then add postgresql/build postgresql/install.

samay-sharma commented:
Pushed a new version

* Removed all copy pasted content
* Addressed feedback (including changing the paths as mentioned)
* Rebased on the latest version

samay-sharma commented:
> > Running tests takes forever - should we have them in short version (current docs don't)?
> 
> Let's discuss that on the list

Left it in and mentioned in the comment that it's optional and takes time.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,147 @@
+  <title>Requirements</title>
+
+  <para>
+   In general, a modern Unix-compatible platform should be able to run
```
+ windows

samay-sharma replied:
Changed.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,147 @@
+  <title>Getting the Source</title>
+
+  <para>
+  You can get the source from github using the instructions below.
```
Hm?

samay-sharma replied:
Was placeholder. Content has been added now.

samay-sharma commented:
> > Where to install and put data directory? (to avoid sudo)
> > I currently put it in the postgres directory. So, the source folder will now have 3 additional folders: build, install and data. This may cause issues with git tracking install and data too. I also don't know if we should assume that we have access / permission to put files in a directory outside the downloaded source. So, want your thoughts on that.
> 
> We could just git clone into postgresql/src and then add postgresql/build postgresql/install.

Done as suggested. There's a working directory called postgres and has src, build, install and data within it.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,147 @@
+   Platforms that are not covered here have no known platform-specific
+   installation issues.
+  </para>
+ </sect1>
```
We should probably add a section comparing make / meson commands to make it easier for people to "migrate". Do you think that's best another section here, or somewhere separate?

samay-sharma replied:
Yes, it makes sense to have such a reference section. Will add.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -387,12 +387,90 @@ meson test
+
+<programlisting>
+git clone git://git.postgresql.org/git/postgresql.git
+</programlisting>
```
I don't think there's much advantage in the git:// protocol left. The only use case I'm aware of is systems that don't have a working https support. I don't think we need to spend effort documenting that.

samay-sharma replied:
Removed.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -387,12 +387,90 @@ meson test
+git fetch
+</programlisting>
+    </para>
+   </step>
```
Hm, that'll not actually update anything. Do we really want to give a git tutorial?

samay-sharma replied:
this is mostly copy pasted, but I should have read it in detail. Will replace with`git pull`.

I think having a line on how to clone a repository and how to fetch recent changes is good enough. I think those 2 are probably required in my opinion and more than that will end up becoming a git tutorial.

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -75,10 +75,316 @@ meson test
+
+  <sect2>
+   <title>Disk space</title>
+   <para>
```
I've updated the values - the old ones were much lower. I think it should be accurate now.

anarazel replied:
It'll very heavily depend on compilation options and compiler.  My build is 350MB, because I enable debug information. And independent of that, it'll just get outdated soon again...

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -75,10 +75,316 @@ meson test
+   default configuration, but they are needed when certain build
+   options are enabled, as explained below:
+
+   <itemizedlist>
```
I already have that section titled Recommended packages. Are there more dependencies you'd like to see moved there?

samay-sharma replied:
Hmm, I will re-order them to put the more important ones before. I think that's the best option, unless you think they qualify for being "Recommended".

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+      </varlistentry>
+
+      <varlistentry>
+       <term><option>-Dkrb-srvnam=<replaceable>NAME</replaceable></option></term>
```
I've lowercased all the - options, as requested by you and Peter Eisentraut.

samay-sharma replied:
Changed. But, if I do meson configure now, I see two options for the ones you've changed (an _ version and a - version)? Is it just my setup or do you see the same?

anarazel replied:
That's likely just due to the old option sticking around in your build directory. A new build dir won't show it, I assume?

samay-sharma replied:
Yes, it doesn't.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+         <literal>DTRACEFLAGS="-64"</literal>.  For example,
+         using the GCC compiler:
+<screen>
+CC='gcc -m64' meson configure -Ddtrace DTRACEFLAGS='-64' ...
```
Don't think this would ever work syntax-wise, and anyway, DTRACEFLAGS isn't implemented.

samay-sharma replied:
Removed section on DTRACEFLAGS

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+<screen>
+CC='gcc -m64' meson configure -Ddtrace DTRACEFLAGS='-64' ...
+</screen>
+         Using Sun's compiler:
```
Sun's compiler isn't supported at the moment, and very unlikely to be supported ever (it's effectively dead, no bugfixes since 2017).

samay-sharma replied:
Removed.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -1104,6 +1104,48 @@ meson configure ... --with-llvm LLVM_CONFIG='/path/to/llvm/bin/llvm-config'
      <variablelist>
 
+      <varlistentry>
+       <term><option>--auto_features=<replaceable>enabled/disabled/auto</replaceable></option></term>
```
I wonder if we should change the enabled/disabled/auto bit in various places to list auto/ first, given it's the default?

samay-sharma replied:
It is in most cases, so makes sense to change the order. However, just wanted to note that it's not the default in all. eg. llvm, dtrace.

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -1104,6 +1104,48 @@ meson configure ... --with-llvm LLVM_CONFIG='/path/to/llvm/bin/llvm-config'
+         The default backend meson uses is ninja and that should suffice for most use cases.
+         However, if you'd like to use a different build system to build Postgres, you can
+         set it using this option. The accepted values for BACKEND are:
+         <command>ninja,vs,vs2010,vs2012,vs2013,vs2015,vs2017,vs2019,vs2022 and xcode</command>.
```
I don't think it's a good idea to list all the vs20* variants here, that'll be guaranteed to be out of date soon. And since xcode isn't known to work... Maybe just add that 'vs' can be used on windows, to fully integrate with visual studio?

eli-schwartz replied:
Meson will anyways detect your VS version and use that, so I'd say it is indeed a good idea to try to elaborate too much here.

anarazel replied:
Can't quite parse your sentence - is there a negation missing?

eli-schwartz replied:
Lol, yeah. :(

samay-sharma replied:
Changed to "The default backend meson uses is ninja and that should suffice for most use cases. However, if you'd like to fully integrate with visual studio, you can set the BACKEND to vs."

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+cd postgres
+
+# fetch source code
+# git clone https://github.com/postgres/postgres.git src
```
Changed to https://git.postgresql.org/git/postgresql.git

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+    <listitem>
+     <para>
+      You can download the source code in two ways - via git or using the source code
+      distribution. For the former, you will need an installed version of
```
changed to downloading the source code tarballs.

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+      be installed with <literal>pip</literal>.
+     </para>
+    </listitem>
+ <listitem>
```
Fixed.

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+
+    <listitem>
+     <para>
+      You need <productname>Zstandard</productname>, if you want to support
```
Added.

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+   <para>
+    The setup command takes a <literal>builddir</literal> and a <literal>srcdir</literal>
+    argument. If no <literal>srcdir</literal> is given Meson will deduce the
+    <literal>srcdir</literal> based on <literal>pwd</literal> and the location
```
Was copied over from meson docs, changed to current directory.

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+
+   <para>
+    Meson then loads the build configuration file and writes the corresponding
+    build backend in the build directory. Additionally, the invocation can pass
```
This is copied over from the meson docs description of [meson setup ](https://mesonbuild.com/Running-Meson.html#configuring-the-build-directory). Do you have a suggested re-write?

anarazel replied:
Ugh, that's awkward phrasing. 'Sets up a build directory' seems sufficient to me?

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+meson setup build --buildtype=debug
+
+#Setup build directory with ssl (Use -D for project specific options)
+meson setup build --Dssl=openssl
```
Yes, this was a typo. Fixed.

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+         use <envar>LLVM_CONFIG</envar> to specify a path to the
+         correct <command>llvm-config</command>. For example
+<programlisting>
+meson configure ... --with-llvm LLVM_CONFIG='/path/to/llvm/bin/llvm-config'
```
I did not and I should have modified it better as this anyway isn't likely to work. Will it work with the below command?

`LLVM_CONFIG=''/path/to/llvm/bin/llvm-config' meson configure ... --with-llvm`

anarazel replied:
As discussed on IM: No, that won't currently help.

eli-schwartz replied:
Meson has some programs it looks up from the environment, see `mesonbuild/envconfig.py` for details.

It will look up things like AR, RANLIB, CC, PKG_CONFIG, CMAKE, VAPIGEN.

It won't look up LLVM_CONFIG, because it's not in that list.

Why is it not in that list? I don't know. Perhaps someone would like to submit a pull request adding it, hint hint. ;)

samay-sharma replied:
@eli-schwartz : Submitted a first draft PR adding it here: https://github.com/mesonbuild/meson/pull/10757 :)

Any tips on what kind of tests would be appropriate?

eli-schwartz replied:
It's using a standard layer in common with other build tools -- and other config-tool dependencies as well -- so I don't think you need to test that it works for llvm specifically.

Thanks! And comments added.

samay-sharma commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1526 @@
+         To use a libxml2 installation that is in an unusual location, you
+         can set <command>pkg-config</command>-related environment
+         variables (see its documentation), or set the environment variable
+         <envar>XML2_CONFIG</envar> to point to
```
Removed the part about on XML2_CONFIG / xml2-config / XML2-CFLAGS etc. Only refer to pkg-config now.

samay-sharma commented:
Addressed all feedback. Things still to discuss: 

* what to do about -Dicu and ICU_FLAGS
* what to do about LLVM_CONFIG

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1708 @@
+       </listitem>
+      </varlistentry>
+
+      <varlistentry>
```
Should we separate the three data layout affecting options into their own list? Then we wouldn't have to repeat the on-disk compatibility thing.

samay-sharma replied:
Done

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1708 @@
+    <para>
+     When building an installation that will be used to develop code inside
+     the server, it is recommended to use at least the
+     options <option>--debug</option>
```
I think we should recommend --buildtype instead of --debug. That's a more general option, and makes it more obvious how to change down to fewer optimizations, which is often needed for debugging.

samay-sharma replied:
Done

anarazel commented on doc/src/sgml/installation-meson.sgml:
```diff
@@ -0,0 +1,1708 @@
+# Run a specific test
+<userinput>meson test recovery/001_stream_rep</userinput>
+
+# Run only the main pg_regress tests
```
That will actually run both pg_regress and isolation tests.

samay-sharma replied:
Corrected.

anarazel commented:
Yeha! Many thanks @samay-sharma
